sydent/threepid/bind.py

```diff
@@ -58,6 +58,9 @@ def addBinding(self, medium: str, address: str, mxid: str) -> Dict[str, Any]:
 
         :return: The signed association.
         """
+        if medium == "email":
+            address = address.casefold()
```

Comment:
We have this sprinkled in a few places.

I can't help but feel that it's a little bit inconsistent that some of these transformations are happening inside the store layer, and some outside.

e.g. `LocalAssociationStore.addOrUpdateAssociation` doesn't transform it, but `GlobalAssociationStore.removeAssociation` does.

~~Have you ever used the library `attrs`? Part of me is thinking that it'd be nice to create a `NormalisedThreepid` class, to store the combination of a `medium` and a **normalised** `address`, though it might be a hassle to thread that through. (The advantage of doing this is that it'd make it difficult to feed in an unnormalised address to code that expects it to be normalised, because type annotations will protect us).~~ Erik says that's overkill :D.

What might be more sensible, though, is to rename things to be either `raw_address` and `normalised_address` and have something like:

`normalised_address = normalise_address(raw_address, medium)`

and have this `normalise_address` function contain the logic to decide what to do to it. This way, if we ever add a new 3PID type that needs a similar processing step, it's just one place that needs changing.
(your migrate script doesn't need to change imo though, since it's already only processing e-mails that unfortunately are already in the database).

Do you think you could do that for the functions in the stores, and the code that you've touched that uses them? (and I suggest moving all the normalisation logic out of the store, just so that there isn't this  potentially-confusing situation where addOrUpdateAssociation expects you to have normalised it yourself, but removeAssociation will do it for you — it seems like the kind of thing that could cause someone to slip up. In other words, the various functions in the stores will be talking about `raw_address`es).

Reply:
Okay so I've been parsing through this and starting at the code and I just want to make sure we are on the same page: the steps suggested are:

Reply:
yup, that's right -- then try and make it clear which functions are using which kind (raw or normalised), at least where you're already touching code (almost 'leave it better than you found it' kind of mentality, if that makes sense) -- but don't worry about combing through the entirety of Sydent, just local improvements where it counts should do :)